### PR TITLE
PROD: Fasten SetMessages::update flags

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Review by Arsnael — APPROVED** [review]

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/mail/InMemoryMessageIdMapper.java [hunk 1]
```diff
@@ -65,6 +68,19 @@ public InMemoryMessageIdMapper(MailboxMapper mailboxMapper, InMemoryMessageMappe
             .block();
     }
 
+    @Override
+    public Publisher<ComposedMessageIdWithMetaData> findMetadata(MessageId messageId) {
+        return mailboxMapper.list()
+            .flatMap(mailbox -> messageMapper.findInMailboxReactive(mailbox, MessageRange.all(), MessageMapper.FetchType.Full, UNLIMITED))
```
Comment (remk) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/mail/InMemoryMessageIdMapper.java:74]: should we have a concurrency value for this flatMap ?

**Review by remk — APPROVED** [review]

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/mail/InMemoryMessageIdMapper.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/mail/InMemoryMessageIdMapper.java:74]: It's in memory, meaning we are not really limited by any backend performance nor risk to overrun stuff.

Overkill IMO

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 2]
```diff
@@ -119,12 +119,20 @@ public StoreMessageIdManager(RightManager rightManager, MailboxSessionMapperFact
     @Override
     public void setFlags(Flags newState, MessageManager.FlagsUpdateMode replace, MessageId messageId, List<MailboxId> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
         MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.getMessageIdMapper(mailboxSession);
+        MailboxMapper mailboxMapper = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession);
+
+        int concurrency = 4;
```
Comment (LanKhuat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java:124]: use a constant?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 3]
```diff
@@ -158,6 +160,17 @@ public void setFlags(Flags newState, MessageManager.FlagsUpdateMode replace, Mes
             .map(Throwing.function(messageResultConverter(fetchGroup)).sneakyThrow());
     }
 
+    @Override
+    public Publisher<ComposedMessageIdWithMetaData> messagesMetadata(Collection<MessageId> ids, MailboxSession session) {
+        MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.getMessageIdMapper(session);
+        int concurrency = 4;
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java:174]: No I prefer to keep it local, as it depict concurrency in this context

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java:124]: idem

**Comment by chibenwa** [comment]
```

[b2b71bc5513a9557d588eb20fb362ffe6641819a] [ERROR]   PostDequeueDecoratorTest.doneShouldCopyMailFromOutboxToSentOnlyOneTimeWhenSuccess:276 » NullPointer
[b2b71bc5513a9557d588eb20fb362ffe6641819a] [ERROR]   PostDequeueDecoratorTest.doneShouldThrowWhenMailboxException »  Unexpected exc...
```

**Review by Arsnael — APPROVED** [review]

**Review by LanKhuat — APPROVED** [review]

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**